### ISSUE TITLE
some questions about the flashattention integerated with megatron

### ISSUE THREAD

**Issue by yingtongxiong** [HazyResearch/flash-attention#257]
Hi, I recently read the megatron code. However, I found that the flashattention integeration was done by author who develop flashattention. I am very confused that why flash attention does not support selective checkpointing? Are there any reasons for doing this?
<img width="504" alt="捕获" src="https://github.com/HazyResearch/flash-attention/assets/45058324/ec0e301b-a1f0-40e6-97b0-adfc98d1ff6e">

Because I recently want to support flashattention with selective checkpointing and met some problems. Therefore, I want to know why don't you support flash-attention with selective checkpointing?

**Comment by ktaebum** [comment]
Memory saving of selective checkpointing is comparable to that of flash attention.
Thus, it is beneficial to use flash attention rather than selective checkpointing if possible because we do not have to "recompute".

**Comment by yingtongxiong** [comment]
You mean that the selective checkpointing has been supported in the flashattention?

**Comment by eggiter** [comment]
@yingtongxiong The FlashAttention algorithm can be seen as a form of `selective gradient checkpointing` because it only store output and statistics tensor other than input activation and recompute  the intermediate values(i.e., attention matrix S&P) in the forward pass. As you can see in the paper:
> we can recompute the attention matrix S and P easily in the backward pass from blocks of Q, K, V in SRAM. This can be seen as a form of selective gradient checkpointing.

**Comment by yingtongxiong** [comment]
you mean that the QK^T output, softmax output and so on don't be stored during forward?

**Comment by eggiter** [comment]
> you mean that the QK^T output, softmax output and so on don't be stored during forward?

Yes, as you can see at line https://github.com/HazyResearch/flash-attention/blob/v1.0.7/flash_attn/flash_attn_interface.py#L63
> ctx.save_for_backward(qkv, out, softmax_lse, cu_seqlens, rng_state)
